### PR TITLE
Add JavaSystemProperties for EO

### PR DESCRIPTION
**Pull request by stanlyDoge**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Enhancement / new feature

### Description
Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/2479

### Checklist

- [ ] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by stanlyDoge** [comment]
UO and TO use different class `EntityOperatorJvmOptions`.

**Comment by scholzj** [comment]
And I was asking why not use the same.

**Comment by stanlyDoge** [comment]
I do not remember exactly but IIRC there was some discussion about setting only GC logging back in https://github.com/strimzi/strimzi-kafka-operator/pull/1143

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests testcase=LogSettingST

File: topic-operator/scripts/topic_operator_run.sh [hunk 1]
```diff
@@ -22,6 +26,14 @@ if [ "$STRIMZI_TLS_ENABLED" = "true" ]; then
...
 fi
 
+if [ -n "$KAFKA_HEAP_OPTS" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} ${KAFKA_HEAP_OPTS}"
+fi
+
+if [ -n "$KAFKA_JVM_PERFORMANCE_OPTS" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} ${KAFKA_JVM_PERFORMANCE_OPTS}"
+fi
+
```
Comment (scholzj) [inline topic-operator/scripts/topic_operator_run.sh]: I do not think we should pull the Kafka env var names here. We should use our own.

File: user-operator/scripts/user_operator_run.sh [hunk 2]
```diff
@@ -3,6 +3,19 @@ if [ -f /opt/user-operator/custom-config/log4j2.properties ];
...
+fi
+
+if [ -n "$KAFKA_HEAP_OPTS" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} ${KAFKA_HEAP_OPTS}"
+fi
+
+if [ -n "$KAFKA_JVM_PERFORMANCE_OPTS" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} ${KAFKA_JVM_PERFORMANCE_OPTS}"
+fi
+
```
Comment (scholzj) [inline user-operator/scripts/user_operator_run.sh]: Same as above

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityTopicOperator.java [hunk 3]
```diff
@@ -259,6 +263,11 @@ public static EntityTopicOperator fromCrd(Kafka kafkaAssembly) {
         varList.add(buildEnvVar(ENV_VAR_TOPIC_METADATA_MAX_ATTEMPTS, String.valueOf(topicMetadataMaxAttempts)));
         varList.add(buildEnvVar(ENV_VAR_TLS_ENABLED, Boolean.toString(true)));
         varList.add(buildEnvVar(ENV_VAR_STRIMZI_GC_LOG_ENABLED, String.valueOf(gcLoggingEnabled)));
+        if (javaSystemProperties != null) {
+            varList.add(buildEnvVar(ENV_VAR_STRIMZI_JAVA_SYSTEM_PROPERTIES, ModelUtils.getJavaSystemPropertiesToString(javaSystemProperties)));
+        }
+        heapOptions(varList, 0.5, 5L * 1024L * 1024L * 1024L);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityTopicOperator.java]: So if i remember correctly, this tries to calculate the defult -Xmx etc.? I do not hink we want that, I think we want to just pass the options over to the containers / JVM.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 4]
```diff
@@ -257,6 +261,11 @@ public static EntityUserOperator fromCrd(Kafka kafkaAssembly) {
         varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_VALIDITY, Integer.toString(clientsCaValidityDays)));
         varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_RENEWAL, Integer.toString(clientsCaRenewalDays)));
         varList.add(buildEnvVar(ENV_VAR_STRIMZI_GC_LOG_ENABLED, String.valueOf(gcLoggingEnabled)));
+        if (javaSystemProperties != null) {
+            varList.add(buildEnvVar(ENV_VAR_STRIMZI_JAVA_SYSTEM_PROPERTIES, ModelUtils.getJavaSystemPropertiesToString(javaSystemProperties)));
+        }
+        heapOptions(varList, 0.5, 5L * 1024L * 1024L * 1024L);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: same as above

File: examples/kafka/kafka-ephemeral.yaml [hunk 5]
```diff
@@ -5,13 +5,13 @@ metadata:
 spec:
   kafka:
     version: 2.4.0
-    replicas: 3
+    replicas: 1
```
Comment (scholzj) [inline examples/kafka/kafka-ephemeral.yaml]: This doesn't look right. I guess you maybe didn't wanted to commit this?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 6]
```diff
@@ -1190,4 +1194,47 @@ protected void addContainerEnvsToExistingEnvs(List<EnvVar> existingEnvs, List<Co
             }
         }
     }
+
+    protected void javaOptionsForEO(List<EnvVar> envVars) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: If it is just for EO, should it be maybe in some EO class and not in the Abstratc Model which is shared for everything?

File: api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-extra-property.yaml [hunk 7]
```diff
@@ -23,7 +23,7 @@ spec:
       -Xmx: "4G"
       -Xms: "2G"
       -XX:
-        -foo: bar
+        foo: bar
```
Comment (scholzj) [inline api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-extra-property.yaml]: So if this was wrong the whole time, why do the tests not crash? Should there been some failing tests or something?

Comment (stanlyDoge) [reply api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-extra-property.yaml]: The test was not wrong... just the property was called `-foo` not `foo` (as it was intended I believe)

File: examples/kafka/kafka-ephemeral.yaml (hunk 5 above)
Comment (stanlyDoge) [reply examples/kafka/kafka-ephemeral.yaml]: Damn, I was sure I reverted changes in this file. Thanks

File: api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-extra-property.yaml (hunk 7 above)
Comment (scholzj) [reply api/src/test/resources/io/strimzi/api/kafka/model/Kafka-with-extra-property.yaml]: Ahh, ok. Uderstand now. Thanks

File: CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 ## 0.18.0
 
+* Add possibility to set Java System Properties for UO and TO via CR yaml
```
Comment (tombentley) [inline CHANGELOG.md]:
```suggestion
* Add possibility to set Java System Properties for User Operator and Topic Operator via `Kafka` CR.
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java [hunk 9]
```diff
@@ -336,4 +338,47 @@ public ServiceAccount generateServiceAccount() {
...
+            xx.forEach((k, v) -> {
+                strimziJavaOpts.append(' ').append("-XX:");
+
+                if ("true".equalsIgnoreCase(v))   {
+                    strimziJavaOpts.append("+").append(k);
+                } else if ("false".equalsIgnoreCase(v)) {
+                    strimziJavaOpts.append("-").append(k);
+                } else  {
+                    strimziJavaOpts.append(k).append("=").append(v);
+                }
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java]: The true/false mapping to +/- seems unintuitive to me. Why not have `JvmOptions.Xx` be a list of XX options. You'd prefix each element  with `-XX:` and just pass it through.

I guess I'd also put the code in ModelUtils, alongside `ModelUtils.getJavaSystemPropertiesToString`. In fact having a `jvmOptionsToString` method in ModelUtils which computes `trim` would be better, I think. Then get rid of this method and let the current caller of this method invoke `jvmOptionsToString` and decide how it gets passed to the container (in particular which env var gets set).

Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java]: > The true/false mapping to +/- seems unintuitive to me. Why not have JvmOptions.Xx be a list of XX options. You'd prefix each element with -XX: and just pass it through.

This already exists for all the Kafka based pods. So we cannot easily change it there. And the whole idea of the PR was to have the same configuration in all places not have different stuff in different pods. So I'm not sure this is worth it.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java]: Well if that's the API we've already adopted, I guess we're stuck with it :frowning_face:.

**Comment by scholzj** [comment]
@tombentley I see that you had comments here and didn't approved this yet. Is this OK to merge for you? Or do you have something more?

**Pull request merged**